apps/theming-designer/src/components/ThemingDesigner.tsx

```diff
@@ -0,0 +1,202 @@
+              <FabricPalette themeRules={this.state.themeRules} />
+              <SemanticSlots />
+            </Stack>
+          </Stack>
+        </Stack>
+      </div>
+    );
+  }
+
+  private _onPrimaryColorPickerChange(newColor: IColor | undefined) {
+    this._onColorChange(this.state.primaryColor, BaseSlots.primaryColor, newColor);
+  }
+
+  private _onTextColorPickerChange(newColor: IColor | undefined) {
+    this._onColorChange(this.state.textColor, BaseSlots.foregroundColor, newColor);
+  }
+
+  private _onBkgColorPickerChange(newColor: IColor | undefined) {
+    this._onColorChange(this.state.backgroundColor, BaseSlots.backgroundColor, newColor);
+  }
+
+  private _makeNewTheme = (): void => {
+    if (this.state.themeRules) {
+      const themeAsJson: { [key: string]: string } = ThemeGenerator.getThemeAsJson(this.state.themeRules);
+
+      const finalTheme = createTheme({
+        ...{ palette: themeAsJson }
+        // isInverted: isDark(this.state.themeRules[BaseSlots[BaseSlots.backgroundColor]].color!)
+      });
+      this.setState({ theme: finalTheme });
+      // const root = document.querySelector('.app') as HTMLElement;
+      // if (root) {
+      //   root.style.backgroundColor = finalTheme.semanticColors.bodyBackground;
+      //   root.style.color = finalTheme.semanticColors.bodyText;
+      // }
+
+      // document.body.style.backgroundColor = finalTheme.semanticColors.bodyBackground;
+      // document.body.style.color = finalTheme.semanticColors.bodyText;
+    }
+  };
+
+  private _onColorChange = (colorToChange: IColor, baseSlot: BaseSlots, newColor: IColor | undefined) => {
+    if (colorToChange === this.state.primaryColor) {
+      this.setState({ primaryColor: newColor! });
+    } else if (colorToChange === this.state.textColor) {
+      this.setState({ textColor: newColor! });
+    } else if (colorToChange === this.state.backgroundColor) {
+      // console.log('got to background', colorToChange);
+      this.setState({ backgroundColor: newColor! });
+    } else {
+      return;
+    }
+    this._async.setTimeout(() => {
+      const themeRules = this.state.themeRules;
+      if (themeRules) {
+        const currentIsDark = isDark(themeRules[BaseSlots[BaseSlots.backgroundColor]].color!);
+        ThemeGenerator.setSlot(themeRules[BaseSlots[baseSlot]], newColor!, currentIsDark, true, true);
+        if (currentIsDark !== isDark(themeRules[BaseSlots[BaseSlots.backgroundColor]].color!)) {
+          // isInverted got swapped, so need to refresh slots with new shading rules
+          ThemeGenerator.insureSlots(themeRules, !currentIsDark);
+        }
+      }
+      this.setState({ themeRules: themeRules }, this._makeNewTheme);
+    }, 20);
+    // 20ms is low enough that you can slowly drag to change color and see that theme,
+    // but high enough that quick changes don't get bogged down by a million changes inbetween
```

Comment:
If you are wanting to only redraw the latest change, you might instead want to use `debounce`, or at least cancel the last invocation of `setState`

Reply:
talked about this offline; went with the colorChangeTimeout approach where clearTimeout's used on the number async.setTimeout returns